subcommand/inject-connect/command.go

```diff
@@ -323,13 +343,68 @@ func (c *Command) Run(args []string) int {
 		TLSConfig: &tls.Config{GetCertificate: c.getCertificate},
 	}
 
-	c.UI.Info(fmt.Sprintf("Listening on %q...", c.flagListen))
-	if err := server.ListenAndServeTLS("", ""); err != nil {
-		c.UI.Error(fmt.Sprintf("Error listening: %s", err))
+	// channel used for health checks
+	// also check to see if we should enable TLS
+	var healthCh chan struct{}
+	tlsEnabled := os.Getenv("CONSUL_CACERT")
+	consulPort := "8500"
+	if tlsEnabled != "" {
+		consulPort = "8501"
+	}
+	// create the health check handler
+	healthcheckHandler := &connectinject.HealthCheckHandler{
+		Log:                hclog.Default().Named("healthCheckHandler"),
+		AclConfig:          api.NamespaceACLConfig{},
+		HFlags:             &flags.HTTPFlags{},
+		Clientset:          c.clientset,
+		ConsulClientScheme: runtime.NewScheme().Name(),
+		ConsulPort:         consulPort,
+	}
+
+	// Build the health check controller and start it
+	healthcheckController := &connectinject.HealthCheckController{
+		Log:        hclog.Default().Named("healthCheckController"),
+		Clientset:  c.clientset,
+		Queue:      nil,
+		Informer:   nil,
+		Handle:     healthcheckHandler,
+		MaxRetries: 10,
+	}
+
+	go func() {
+		// TODO error handling to close this
+		c.UI.Info(fmt.Sprintf("Listening on %q...", c.flagListen))
+		if err := server.ListenAndServeTLS("", ""); err != nil {
+			c.UI.Error(fmt.Sprintf("Error listening: %s", err))
+		}
+	}()
+
+	// Start the health check controller
+	// First run Reconcile then start the watch
+	healthCh = make(chan struct{})
+	go func() {
+		defer close(healthCh)
+		healthcheckController.Init(ctx.Done())
+		healthcheckController.Run(ctx.Done())
```

Comment:
Is there a race condition here if a pod changes status after `Init()` has started but before `Run()` is run?

Reply:
This has been addressed via the periodic reconcile loop!